content/blog/2019-10-23-AutoVectoring/index.md

```diff
@@ -0,0 +1,242 @@
+
+For example, printing the *i* each iteration, where *i* is the induction variable, should still allow the loop to be vectorized as there are no flow-dependencies, but if we change the induction variable by four each iteration, that print statment will behave incorrectly as it would only print one out of four elements. To mitigate this, we do partial loop unrolling when strip mining.
+
+To achieve this, we go through the loop and find non-array instructions and keep track of them. Then, we append them to the end of the loop (before the condition and branch statements) and also a copy of the induction variable increment/decrement. We do this insertion three times total. 
+
+Example loop snippit before strip mining and partial unrolling:
+```gherkin=
+loop:
+  print i;
+  ci: int = add c i;
+  v: int = lw ci;
+  i: int = add i inc;
+  done: bool = eq i size;
+  br done exit loop;
+```
+After strip mining and partial unrolling:
+```gherkin=
+loop:
+  print i;
+  ci: int = add c i;
+  v: int = vload ci;
+  i: int = add i inc;
+  print i;
+  i: int = add i inc;
+  print i;
+  i: int = add i inc;
+  print i;
+  i: int = add i inc;
+  done: bool = eq i size;
+  br done exit loop;
+```
+
+With this method, array operations are allowed to be vectorized while preserving serial instructions because serial instructions are unrolled into four copies, where each copy operates with a different induction variable value. This loop also increments the induction variable by four every iteration which preserves the performance increase from reduced branch overhead. 
+
+In terms of implentaion, we found it much easier to first coalesce the loop into one big block before running the strip mine algorithm because we could treat this sequence of instructions as one array instead of having to worry about jumps and label renaming from inserting new blocks. Aggregating all the blocks of this loop was possible because we previously enforced that there can be exactly one branch instruction located at the end of the loop.
+
+Up to here, we have been operating on the assumption that array sizes are divisible by the vector size---four. To account for arrays not divisible by four, we append a copy of this loop (without any optimizations) to a block that follows the main optimized loop. This serially executes the remaining loop iteration which allows us to maintain correctness.
+
+
+### Rust Dynamic Library with Foreign Function Interface
+In order to utilize SIMD intrinsics, we use the Rust crate for SIMD that targets the x86 platform. We design a dynamic library with functions for each of the SIMD operations we support, namely vector-add, vector-multiply, and vector-subtract. These functions are called in the dispatch loop of the interpreter for the corresponding vectorized instructions.
```

Comment:
Wow! That's quite a thing. I think you could have gotten away with "fake" vectorization in the interpreter, but this is even better.